### PR TITLE
If we are building mlpack statically then build executable statically

### PR DESCRIPTION
**Pull request by shrit**
This pull request is not related to cross-compilation.

Signed-off-by: Omar Shrit <omar@shrit.me>

### CODE REVIEW

**Review by rcurtin — COMMENTED** [review]
Can you also make sure the documentation is updated? :+1:

**Comment by shrit** [comment]
Do you have a specific file in mind to update? for example `doc/guide/build.hpp`?

**Comment by rcurtin** [comment]
The original PR had some modifications to the documentation of the options in `CMakeLists.txt`.  And yeah, we probably should also update the descriptions in `README.md` and `doc/guide/build.hpp` too. :+1:

**Comment by shrit** [comment]
Exactly, the modifications related to the main `CMakeLists.txt` are already embedded in the autodownloader pull request #2927, sorry for that. 
I will add additional one in `README.md` and in `doc/guide/build.hpp`

File: doc/guide/build.hpp [hunk 1]
```diff
@@ -209,17 +209,27 @@ The full list of options mlpack allows:
 Each option can be specified to CMake with the '-D' flag.  Other tools can also
 be used to configure CMake, but those are not documented here.
 
+For example, if you would like to build mlpack and its CLI binding statically, then
```
Comment (rcurtin) [inline doc/guide/build.hpp:212]:
```suggestion
For example, if you would like to build mlpack and its CLI bindings statically, then
```
Minor pedantry. :)

Comment (rcurtin) [reply doc/guide/build.hpp:212]: Also, do you think you can update the `BUILD_SHARED_LIBS` documentation above?  Right now it just mentions libraries, but it should be "libraries and executables".

File: src/mlpack/bindings/cli/CMakeLists.txt [hunk 2]
```diff
@@ -50,11 +50,21 @@ if (BUILD_CLI_EXECUTABLES)
...
-    mlpack
-    ${ARMADILLO_LIBRARIES}
-    ${COMPILER_SUPPORT_LIBRARIES}
-  )
+  # Build mlpack CLI binding binaries statically.
+  if(NOT BUILD_SHARED_LIBS)
+    target_link_libraries(mlpack_${name} -static
+     mlpack
+     ${ARMADILLO_LIBRARIES}
+     ${COMPILER_SUPPORT_LIBRARIES}
```
Comment (rcurtin) [inline src/mlpack/bindings/cli/CMakeLists.txt:58]:
```suggestion
      mlpack
      ${ARMADILLO_LIBRARIES}
      ${COMPILER_SUPPORT_LIBRARIES}
```
Tiny style fix. :)

**Review by rcurtin — COMMENTED** [review]
Awesome, thanks---this looks good to me if you can fix the couple tiny comments. :)  Thanks for adding `CEREAL_INCLUDE_DIR` to the documentation also!

**Review by rcurtin — COMMENTED** [review]
Also did you want to modify the README to discuss `BUILD_SHARED_LIBS` too?

File: doc/guide/build.hpp (hunk 1 above)
Comment (shrit) [reply doc/guide/build.hpp:212]: Yeah, of course

**Comment by shrit** [comment]
@rcurtin, sorry I forget about the readme, I will add one line to declare the option :+1:

File: README.md [hunk 3]
```diff
@@ -211,7 +211,7 @@ Options are specified with the -D flag.  The allowed options include:
     BUILD_R_BINDINGS=(ON/OFF): whether or not to build R bindings
     R_EXECUTABLE=(/path/to/R): Path to specific R executable
     BUILD_TESTS=(ON/OFF): whether or not to build tests
-    BUILD_SHARED_LIBS=(ON/OFF): compile shared libraries as opposed to
+    BUILD_SHARED_LIBS=(ON/OFF): compile shared libraries and executables as opposed to
        static libraries
```
Comment (rcurtin) [inline README.md:215]:
```suggestion
    BUILD_SHARED_LIBS=(ON/OFF): compile shared libraries and executables as
        opposed to static libraries
```
Oops, now the line is too long. :smile:

File: src/mlpack/bindings/cli/CMakeLists.txt [hunk 4]
```diff
@@ -50,11 +50,21 @@ if (BUILD_CLI_EXECUTABLES)
...
-    ${ARMADILLO_LIBRARIES}
-    ${COMPILER_SUPPORT_LIBRARIES}
-  )
+  # Build mlpack CLI binding binaries statically.
+  if(NOT BUILD_SHARED_LIBS)
+    target_link_libraries(mlpack_${name} -static
+      mlpack
+      ${ARMADILLO_LIBRARIES}
+      ${COMPILER_SUPPORT_LIBRARIES}
+   )
```
Comment (rcurtin) [inline src/mlpack/bindings/cli/CMakeLists.txt:59]:
```suggestion
    )
```
Ah, missed this one last time.

**Review by rcurtin — APPROVED** [review]
Awesome, looks good to me.  Just two more tiny changes. :)

**Comment by rcurtin** [comment]
Awesome, thanks! :+1:

**Comment by shrit** [comment]
Your are most welcome, I will open one tomorrow for the embedded system :+1:

**Pull request merged**